d3a_api_client/rest_device.py

```diff
@@ -139,28 +143,40 @@ def _on_market_cycle(self, message):
         logging.debug(f"A new market was created. Market information: {message}")
 
         def executor_function():
-            self.on_market_cycle(message)
+            try:
+                self.on_market_cycle(message)
+            except Exception as e:
+                logging.error(f"on_market_cycle raised exception: {e}. Traceback: {traceback.format_exc()}")
         self.callback_thread.submit(executor_function)
 
     def _on_tick(self, message):
         logging.debug(f"Time has elapsed on the device. Progress info: {message}")
 
         def executor_function():
-            self.on_tick(message)
+            try:
+                self.on_tick(message)
+            except Exception as e:
+                logging.error(f"on_tick raised exception: {e}. Traceback: {traceback.format_exc()}")
         self.callback_thread.submit(executor_function)
 
     def _on_trade(self, message):
         logging.debug(f"A trade took place on the device. Trade information: {message}")
 
         def executor_function():
-            self.on_trade(message)
+            try:
+                self.on_trade(message)
+            except Exception as e:
+                logging.error(f"on_trade raised exception: {e}. Traceback: {traceback.format_exc()}")
         self.callback_thread.submit(executor_function)
 
     def _on_finish(self, message):
         logging.debug(f"Simulation finished. Information: {message}")
 
         def executor_function():
-            self.on_finish(message)
+            try:
+                self.on_finish(message)
+            except Exception as e:
+                logging.error(f"on_finish raised exception: {e}. Traceback: {traceback.format_exc()}")
```

Comment:
In order to make it more straightforward to detect which device failed in multi-device tests, it would be great to also have the device_id in the log too.

Reply:
Good idea! Will add it. Thanks